### ISSUE TITLE
[BUG] starlette 0.21.0 uses httpx testclient

### ISSUE THREAD

**Issue by grigi** [schemathesis/schemathesis#1637]
**Checklist**

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq.html#frequently-asked-questions) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)

**Describe the bug**
`starlette==0.21.0` updated to use `httpx` for the testclient instead of `requests`.

In particular `starlette.testclient.TestClient` now inherits `httpx.Client`.

**To Reproduce**
Use `Case.call_asgi(...)` in a unit test.

Error:
```
  File "/mnt/win/work/nightcrawler/tests/schema/lookup/test_txt.py", line 22, in test_txt
    response = case.call_asgi(headers={'Authorization': 'Bearer supersecret'})
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/schemathesis/models.py", line 397, in call_asgi
    return self.call(base_url=base_url, session=client, headers=headers, **kwargs)
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/schemathesis/models.py", line 327, in call
    response = session.request(**data)  # type: ignore
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/starlette/testclient.py", line 448, in request
    return super().request(
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/httpx/_client.py", line 803, in request
    request = self.build_request(
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/httpx/_client.py", line 346, in build_request
    headers = self._merge_headers(headers)
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/httpx/_client.py", line 413, in _merge_headers
    merged_headers.update(headers)
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/httpx/_models.py", line 199, in update
    headers = Headers(headers)
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/httpx/_models.py", line 79, in __init__
    self._list = [
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/httpx/_models.py", line 85, in <listcomp>
    for k, v in headers
ValueError: too many values to unpack (expected 2)
Falsifying example: test_txt(
    case=Case(path_parameters={'domain': 'A.ac'}, headers={'Authorization': 'Bearer '}, query={}),
    self=<tests.schema.lookup.test_txt.TestTXTSchema testMethod=test_txt>,
)
```

**Additional context**
I forced the headers to be a dict instead of a `requests.structures.CaseInsensitiveDict` and that furthered the state, but it next up failed with:
```
  File "/mnt/win/work/nightcrawler/tests/schema/lookup/test_txt.py", line 23, in test_txt
    case.validate_response(response)
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/schemathesis/models.py", line 424, in validate_response
    copied_response = copy_response(response)
  File "/mnt/win/work/nightcrawler/.pyenv/lib/python3.10/site-packages/schemathesis/utils.py", line 279, in copy_response
    response.freeze()
AttributeError: 'Response' object has no attribute 'freeze'
Falsifying example: test_txt(
    case=Case(path_parameters={'domain': 'A.ac'}, headers={'Authorization': 'Bearer '}, query={}),
    self=<tests.schema.lookup.test_txt.TestTXTSchema testMethod=test_txt>,
)
```
This seems to be caused by now getting a `httpx.Response` back instead of a `requests.Response`, and these objects are very different.

**Comment by Stranger6667** [comment]
Oh, right. Thank you for opening this issue. I'll take a look this evening

**Comment by Stranger6667** [comment]
This will take some more time as it requires full support for `httpx`, so all built-in checks, hooks, the runner integration, etc should be updated.

**Comment by hillairet** [comment]
As a consequence `FastAPI >= 0.87.0` is also affected since it uses `Starlette >= 0.21.0`.
So thank you in advance for handling `httpx`. :pray:

**Comment by dkbarn** [comment]
@Stranger6667 if it is going to take a while to address this issue, in the meantime please can the `pyproject.toml` be corrected to specify its dependency on the appropriate version of starlette?

https://github.com/schemathesis/schemathesis/blob/master/pyproject.toml#L52
at the moment this is specified as `starlette = ">=0.13,<1"` which is incorrect, since schemathesis is not compatible with starlette>0.21

**Comment by Kludex** [comment]
@Stranger6667 You can ping me on those issues, I can help.

Please do not restrict the range of versions of Starlette. You can replace the Starlette `TestClient` by this one: https://github.com/Kludex/starlette-testclient; and I've also created this tool to help on the bump: https://github.com/Kludex/bump-testclient.

**Comment by grigi** [comment]
I think using the lib that Kludex linked to would resolve the issue in the interm, and can be removed when a proper httpx testclient is completed?

**Comment by Stranger6667** [comment]
Sorry for the delay!

I ran with the approach suggested by @Kludex and it will be available in the next version soon. However, I'd like to overhaul transports properly as it unblocks a few nice things (like a unified interface to requests/WSGI/ASGI responses and adding more transports in the future, etc)

**Comment by tmeckel** [comment]
@Stranger6667 Sorry, does not work for me.

![image](https://user-images.githubusercontent.com/14177833/216567228-51b99b3b-205d-405d-9443-9268655e3c0d.png)

![image](https://user-images.githubusercontent.com/14177833/216567269-8e40bf5e-e4e7-49cd-ab02-1a355d631c03.png)

![image](https://user-images.githubusercontent.com/14177833/216567346-64c41084-62b0-4c76-a716-0d3f5eba55c6.png)

**Comment by Stranger6667** [comment]
@tmeckel Could you, please, share the `test_api` function definition + how you load your schema (`schemathesis.from_*` call)?

**Comment by tmeckel** [comment]
@Stranger6667 Okay, I suspect, because I use the FastAPI TestClient, it's internally bound to the "nomal" `starlette`  client and not to the client from the `starlette_client` package from @Kludex 

![image](https://user-images.githubusercontent.com/14177833/216592742-4504199f-689b-4499-8e16-13f15133dced.png)

Any thing I can do about it?

**Comment by Stranger6667** [comment]
I suggest passing the `app` fixture to `schemathesis.from_uri` instead of `client`. Then it will be wrapped with `starlette_client.TestClient` internally. Let me know if it works for you

**Comment by tmeckel** [comment]
@Stranger6667 No, this didn't work. 

![image](https://user-images.githubusercontent.com/14177833/216771880-1cc3811a-134b-4a60-91db-f7e0afeb2288.png)

```python
client = <starlette.testclient.TestClient object at 0x7f7628538be0>

    @pytest.mark.skipif(
>       parse_version(schemathesis.__version__) < parse_version("3.18.0")
        and parse_version(starlette.__version__) >= parse_version("0.21.0"),
        reason=f"schemathesis <= {schemathesis.__version__} does not currently support starlette >= 0.21.0 https://github.com/schemathesis/schemathesis/issues/1637",
    )

tests/test_schemathesis.py:97: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/test_schemathesis.py:105: in test_api
    case.call_and_validate(session=client)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

response = <Response [500 Internal Server Error]>

    def copy_response(response: GenericResponse) -> GenericResponse:
        """Create a copy of the given response as far as it makes sense."""
        if isinstance(response, requests.Response):
            # Hooks are not copyable. Keep them out and copy the rest
            hooks = None
            if response.request is not None:
                hooks = response.request.hooks["response"]
                response.request.hooks["response"] = []
            copied_response = deepcopy(response)
            if hooks is not None:
                copied_response.request.hooks["response"] = hooks
            setattr(copied_response, "raw", response.raw)
            return copied_response
        # Can't deepcopy WSGI response due to generators inside (`response.freeze` doesn't completely help)
>       response.freeze()
E       AttributeError: 'Response' object has no attribute 'freeze'
E       Falsifying example: test_api(
E           case=Case(path_parameters={'id': '0'}, body=[]),
E           client=<starlette.testclient.TestClient object at 0x7f7628538be0>,
E       )

.venv/lib/python3.9/site-packages/schemathesis/utils.py:279: AttributeError
```

I even tried to pass the `app` instead of the `client` fixture to `test_api`, but this created a different error:

```python
tests/test_schemathesis.py:105: in test_api
    case.call_and_validate(session=app)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Case(path_parameters={'id': '0'}), base_url = None, session = <fastapi.applications.FastAPI object at 0x7f73d7120100>, headers = None, kwargs = {}
hook_context = HookContext(operation=APIOperation(path='/deployments/{id}/outputs', method='delete', schema=OpenApi30 for UI Cloud Pl...ms=[]), query=ParameterSet(items=[]), body=PayloadAlternatives(items=[]), case_cls=<class 'schemathesis.models.Case'>))
data = {'cookies': None, 'headers': {'User-Agent': 'schemathesis/3.18.0', 'X-Schemathesis-TestCaseId': '02adb61b4a184036bc9cd40c987999a5'}, 'method': 'DELETE', 'params': None, ...}, close_session = False

    def call(
        self,
        base_url: Optional[str] = None,
        session: Optional[requests.Session] = None,
        headers: Optional[Dict[str, Any]] = None,
        **kwargs: Any,
    ) -> requests.Response:
        """Make a network call with `requests`."""
        hook_context = HookContext(operation=self.operation)
        dispatch("before_call", hook_context, self)
        data = self.as_requests_kwargs(base_url, headers)
        data.update(kwargs)
        data.setdefault("timeout", DEFAULT_RESPONSE_TIMEOUT / 1000)
        if session is None:
            validate_vanilla_requests_kwargs(data)
            session = requests.Session()
            close_session = True
        else:
            close_session = False
        try:
>           response = session.request(**data)  # type: ignore
E           AttributeError: 'FastAPI' object has no attribute 'request'
E           Falsifying example: test_api(
E               case=Case(path_parameters={'id': '0'}),
E               app=<fastapi.applications.FastAPI object at 0x7f73d7120100>,
E           )

.venv/lib/python3.9/site-packages/schemathesis/models.py:358: AttributeError
```

Perhaps I'm using `schemathesis` completely wrong, but until `starlette` changed to `httpx`, my code ran perfectly and helped me to discover a bunch of nasty bugs.

**Comment by Stranger6667** [comment]
Thank you for sharing the code! I believe you need to pass `starlette_client.TestClient` instance on line 105, instead of the one from `starlette`

**Comment by tmeckel** [comment]
@Stranger6667 Dammit! So obvious!!! This solved the issue. Thanks for helping!

![image](https://user-images.githubusercontent.com/14177833/216773075-ab038741-4ed4-4c26-b5bb-2d9e03fb2d70.png)

**Comment by Stranger6667** [comment]
@tmeckel You're very welcome! :)

**Comment by thatguysimon** [comment]
Hey @Stranger6667!
The fact that this issue is closed gives the false impression that it was fixed and can accidentally send folks (like myself) on a wild goose chase 😄 
The dependency conflict can indeed be resolved but Schemathesis still doesn't support `httpx` (@Kludex's [workaround](https://github.com/schemathesis/schemathesis/issues/1637#issuecomment-1364661574) is a lifesaver in the meantime 🙏🏻 ).

Is it possible to reopen this issue so this is communicated better? 
Or alternatively, should I open a new issue for this?

Thanks!

**Comment by Stranger6667** [comment]
Hi @thatguysimon !

Sorry, I forgot to open a new issue - I wanted to dig into more details about what is needed for proper `httpx` support. I'd appreciate it if you could open a new issue :) I'll fill in more information there once I have more bandwidth

**Comment by thatguysimon** [comment]
@Stranger6667 Opened https://github.com/schemathesis/schemathesis/issues/1718
I'm available to take some dev tasks if you need some help by the way.
